### PR TITLE
Merge sequence-model into main

### PR DESCRIPTION
**Pull request by dolevi-tub**
This PR adds a basic implementation of templating and code generation for python3.
It also adds a happy case test for the code generation as well as test execution in github actions for the project.

### CODE REVIEW

File: src/app/transformation/test_template.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+              "\n"
+              "def run(dependee):\n"
+              "    dependee.function(ins_param1=5, ins_param2='5')\n"
+              "    dependee.method(ins_param2='5', ins_param1=5)\n"
+              "\n"
+              "\n"
+              "if __name__ == '__main__':\n"
+              "    init()\n"
+              "\n"
+              "    dependee = 5\n"
```
Comment (janosbrodbeck) [inline src/app/transformation/test_template.py:52]: Right now, we have the possibility to have dependees for numbers or simple strings. When we look at our first example [code](https://raw.githubusercontent.com/SQE22-RoboticWelding/deployment_examples/examples_janos/sample_welding_sequence.py) we see, that we need `n = NiryoOne()`. Currently, we would do that by having a dependee named `n` and with the initializer set to `NiryoOne()`, as string. 

In our example code we called `NiryoOne()` without parameters as we used the defaults. Normally you would also provide an IP and with a good API also the port....but that is hardcoded in the Niryo API atm  ¯\\_(ツ)_/¯.  
So we basically initialize dependees with a hardcoded statement and do not support dependees, which need multiple parameters like `n = NiryoOne("10.0.0.1", 8080)` for initialization itself. Sadly, this kind of means dependees for dependees, but I feel that we need it at some point.

Comment (janosbrodbeck) [reply src/app/transformation/test_template.py:52]: But I am totally fine if we tackle that on a later point.

File: src/app/transformation/abstract_model.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        self.abstract_name = abstract_name
+        self.generated_name = generated_name
+        self.type = type
+        self.description = description
+
+
+class Action:
+    abstract_name: str
+    generated_name: str
+    generated_actor: DependeeReference
```
Comment (janosbrodbeck) [inline src/app/transformation/abstract_model.py:65]: I know why you designed it like that, but I think we limit us a bit too much, when we only allow actions which have a predefined dependee. I'm thinking about calling static functions of a class from a library, e.g. `NiryoOne.Reset()` (completely made up). I would totally ignore that problem for now, but we should maybe investigate whether we need support for such calls.

I think we should allow actor to be empty in the long run to allow function calls without having an actor. But that would then also mean that the template needs to support it (see comment in the template. But I am also fine with tackling that later).

**Review by janosbrodbeck — CHANGES_REQUESTED** [review]
Awesome work overall. The naming is not quite intuitive at first, but that is primarily due to the two levels of abstraction and in general being quite generic. I think this is difficult to avoid. I can also not yet fully judge how difficult it will be to provide the user a UI for creating or edit a template in an easy way. 

My comments are primarily things to allow more language features. I don't think we can fully estimate what we need to support already, so I would prefer to address that in further iterations and not directly everything in this PR. Only my comment with the return in `init` I see as "should be fixed". 

Otherwise, looks very good to me!

File: src/app/transformation/test_template.py (hunk 1 above)
Comment (dolevi-tub) [reply src/app/transformation/test_template.py:52]: It is possible to provide paramters to an `Dependee.initializer: ProgrammingStatement` like this:
```python3
from app.transformation.abstract_model import Dependee

d = Dependee("niryo", "NiryoOne('10.0.0.1', 8080)")
```
Or do you mean a construct like this (emphasis on the `c`)?
```python3
from app.transformation.abstract_model import Dependee

c = Dependee("niryo", "NiryoOne()")
d = Dependee("other_niryo", "NiryoOne(c)")
```

I would also consider to just let these (even more elaborate) setup statements be written by the template creator. The `string` type doesn't enforce any level of sophistication and neither do I think we must/should.

Comment (janosbrodbeck) [reply src/app/transformation/test_template.py]: The `ProgrammingStatement` is still just a string. For `d = Dependee("niryo", "NiryoOne('10.0.0.1', 8080)")` we would have the Dependee with the initializer set to `"niryo", "NiryoOne('10.0.0.1', 8080)"`. I would not see the exact port and IP as part of the template, as these parameters can change depending on configuration and respective used robot. Robot A could be reachable at `10.0.0.1`, Robot B `10.0.0.2`. 
But we don't know exactly how flexible we have to be here yet. Its just an assumption from me that we will need this.

**Review by janosbrodbeck — APPROVED** [review]

**Pull request merged**